## Data source:
https://xkcd.com/json.html

In [1]:
import pandas as pd
import os
import time

## This section will download a dataframe of all the xkcd comics from 0-newest

In [2]:
# doesn't need to be run 
# but helps to check that everything is up
test = pd.read_json("https://xkcd.com/614/info.0.json", typ = "series")
xkcdDf = pd.DataFrame(columns = test.index)
xkcdDf.loc[0] = test
xkcdDf

month  num link  year news  safe_title  \
0     7  614       2009       Woodpecker   

                                          transcript  \
0  [[A man with a beret and a woman are standing ...   

                                                 alt  \
0  If you don't have an extension cord I can get ...   

                                           img       title day  
0  https://imgs.xkcd.com/comics/woodpecker.png  Woodpecker  24

In [3]:
%%time
#pulls the newest comic
newestXkcd = pd.read_json("https://xkcd.com/info.0.json" , typ = "series")
xkcdDf = pd.DataFrame(columns = newestXkcd.index)
#gets the number of the newest comic so we can work backwards from there
newestNumber = newestXkcd["num"]

errorIssues = []
for issue in range(newestNumber):
    try:
        currentLink = "https://xkcd.com/{num}/info.0.json".format(num = issue)
        xkcdDf.loc[issue] = pd.read_json(currentLink , typ = "series")
    except:
        errorIssues += [issue]
        
    print("\r Currently on Issue#: {0}".format(issue), end='')
    time.sleep(.1)

print("")    
xkcdDf

 Currently on Issue#: 2441Wall time: 13min 6s


month   num link  year news              safe_title  \
1        1     1       2006              Barrel - Part 1   
2        1     2       2006         Petit Trees (sketch)   
3        1     3       2006              Island (sketch)   
4        1     4       2006           Landscape (sketch)   
5        1     5       2006                  Blown apart   
...    ...   ...  ...   ...  ...                     ...   
2437     3  2437       2021           Post-Vaccine Party   
2438     3  2438       2021                         Siri   
2439     3  2439       2021       Solar System Cartogram   
2440     3  2440       2021        Epistemic Uncertainty   
2441     3  2441       2021                IMDb Vaccines   

                                             transcript  \
1     [[A boy sits in a barrel which is floating in ...   
2     [[Two trees are growing on opposite sides of a...   
3      [[A sketch of an Island]]\n{{Alt:Hello, island}}   
4     [[A sketch of a landscape with sun on the hori...   
5     [[A black number 70 sees a red package.]]\n70:...   
...                                                 ...   
2437                                                      
2438                                                      
2439                                                      
2440                                                      
2441                                                      

                                                    alt  \
1                                         Don't we all.   
2     'Petit' being a reference to Le Petit Prince, ...   
3                                         Hello, island   
4             There's a river flowing through the ocean   
5                              Blown into prime factors   
...                                                 ...   
2437  [Future update] Well, someone accidentally dro...   
2438  Alexa defeated her in a battle hinging on the ...   
2439  For sentimental reasons, every active Mars rov...   
2440  Luckily, unlike in our previous study, we have...   
2441  I'm actually not sure if Vader and the Emperor...   

                                                    img  \
1     https://imgs.xkcd.com/comics/barrel_cropped_(1...   
2     https://imgs.xkcd.com/comics/tree_cropped_(1).jpg   
3         https://imgs.xkcd.com/comics/island_color.jpg   
4     https://imgs.xkcd.com/comics/landscape_cropped...   
5     https://imgs.xkcd.com/comics/blownapart_color.jpg   
...                                                 ...   
2437  https://imgs.xkcd.com/comics/post_vaccine_part...   
2438              https://imgs.xkcd.com/comics/siri.png   
2439  https://imgs.xkcd.com/comics/solar_system_cart...   
2440  https://imgs.xkcd.com/comics/epistemic_uncerta...   
2441     https://imgs.xkcd.com/comics/imdb_vaccines.png   

                       title day  
1            Barrel - Part 1   1  
2       Petit Trees (sketch)   1  
3            Island (sketch)   1  
4         Landscape (sketch)   1  
5                Blown apart   1  
...                      ...  ..  
2437      Post-Vaccine Party  15  
2438                    Siri  17  
2439  Solar System Cartogram  19  
2440   Epistemic Uncertainty  22  
2441           IMDb Vaccines  24  

[2440 rows x 11 columns]

In [4]:
# Comics that didn't download properly
print("Missing comic #'s",errorIssues,"", sep = '\n')

Missing comic #'s
[0, 404]



Notice there is no comic #0 and #404, 404 is likely a joke because of the common error for missingness.  
I thought there would be one for 0 since they have tons of coding related humor, but guess not  
#1965 is strange error I encountered, this never happened before. 
After investigating, it worked perfectly fine by going to the json with the same link manually. Strange

In [5]:
# number of comics in df (including errors)
print('Newest comic number', newestNumber)
print('Total Comics: ', newestNumber-len(errorIssues)+1) # +1 because of them not starting the index at 0

Newest comic number 2442
Total Comics:  2441


## Now making a system that UPDATES the csv rather than starting fresh every time

In [6]:
#finding the newest dataframe

datas = os.listdir('data')
print("files in data directory:", datas,"", sep = '\n')
if len(datas) == 0:
    currentNum = 1
else:
    tableNumbers = [int(val[10:13]) for val in datas]
    print('Table nums that exist', tableNumbers)
    currentNum = max(tableNumbers) + 1
print('num for the next table: ', currentNum)

files in data directory:
['xkcd_data_001.csv', 'xkcd_data_002.csv', 'xkcd_data_003.csv', 'xkcd_data_004.csv']

Table nums that exist [1, 2, 3, 4]
num for the next table:  5


### ONLY run this if data is empty or needs to use the data from the section above

In [7]:
outfp = os.path.join('data', 'xkcd_data_' + str(currentNum).zfill(3) + '.csv')
print('path to next file: ', outfp)
xkcdDf.to_csv(outfp,index = False)

path to next file:  data\xkcd_data_005.csv


## Now for updating

In [8]:
#do this next time when there is more data
infp = os.path.join('data', 'xkcd_data_' + str(currentNum-1).zfill(3) + '.csv')
pd.read_csv(infp)

month  num  link  year  news                 safe_title  \
0       1    1   NaN  2006   NaN            Barrel - Part 1   
1       1    2   NaN  2006   NaN       Petit Trees (sketch)   
2       1    3   NaN  2006   NaN            Island (sketch)   
3       1    4   NaN  2006   NaN         Landscape (sketch)   
4       1    5   NaN  2006   NaN                Blown apart   
..    ...  ...   ...   ...   ...                        ...   
79      3   82   NaN  2006   NaN                      Frame   
80      3   83   NaN  2006   NaN                   Katamari   
81      4   84   NaN  2006   NaN          National Language   
82      4   85   NaN  2006   NaN                      Paths   
83      4   86   NaN  2006   NaN  Digital Rights Management   

                                           transcript  \
0   [[A boy sits in a barrel which is floating in ...   
1   [[Two trees are growing on opposite sides of a...   
2    [[A sketch of an Island]]\n{{Alt:Hello, island}}   
3   [[A sketch of a landscape with sun on the hori...   
4   [[A black number 70 sees a red package.]]\n70:...   
..                                                ...   
79  [[A stick figure stands alone in the centre of...   
80  [[A girl stands on the left.  A man is sitting...   
81  This happened to my friend:\n[[Men and women a...   
82  [[Blueprint of a campus. Two buildings in the ...   
83  [[Hat Guy is standing on an advancing glacier]...   

                                                  alt  \
0                                       Don't we all.   
1   'Petit' being a reference to Le Petit Prince, ...   
2                                       Hello, island   
3           There's a river flowing through the ocean   
4                            Blown into prime factors   
..                                                ...   
79                                                ...   
80  As the King of All Cosmos remarked, 'Is it tha...   
81                  She's pretty sharp when provoked.   
82        It's true, I think about this all the time.   
83  If you're interested in the subject, Lawrence ...   

                                                  img  \
0   https://imgs.xkcd.com/comics/barrel_cropped_(1...   
1   https://imgs.xkcd.com/comics/tree_cropped_(1).jpg   
2       https://imgs.xkcd.com/comics/island_color.jpg   
3   https://imgs.xkcd.com/comics/landscape_cropped...   
4   https://imgs.xkcd.com/comics/blownapart_color.jpg   
..                                                ...   
79             https://imgs.xkcd.com/comics/frame.jpg   
80          https://imgs.xkcd.com/comics/katamari.jpg   
81  https://imgs.xkcd.com/comics/national_language...   
82             https://imgs.xkcd.com/comics/paths.jpg   
83  https://imgs.xkcd.com/comics/sony_microsoft_mp...   

                        title  day  
0             Barrel - Part 1    1  
1        Petit Trees (sketch)    1  
2             Island (sketch)    1  
3          Landscape (sketch)    1  
4                 Blown apart    1  
..                        ...  ...  
79                      Frame   29  
80                   Katamari   31  
81          National Language    3  
82                      Paths    5  
83  Digital Rights Management    7  

[84 rows x 11 columns]